### PR TITLE
Add optional Fluentd support

### PR DESCRIPTION
**Pull request by femueller**
Signed-off-by: Felix Mueller <felix.mueller@zalando.de>

This PR will add support for Fluentd as an alternative distributed logging agent.

Backward compatible, the Scalyr configuration up until now is still recognized and Scalyr agent will still be used, Fluentd will not be run in that case.

With Fluentd the following destinations for logs are available as of now: scalyr, s3, rsyslog

### CODE REVIEW

File: build/setup.d/21-install-td-agent-packages.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Ensure that /etc/init.d/td-agent is removed
+rm /etc/init.d/td-agent
+
+# Ensure that directories are writable
+chmod -R 1777 /tmp/
```
Comment (aermakov-zalando) [inline build/setup.d/21-install-td-agent-packages.sh]: Why is this needed?

File: build/setup.d/21-install-td-agent-packages.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Ensure that /etc/init.d/td-agent is removed
```
Comment (aermakov-zalando) [inline build/setup.d/21-install-td-agent-packages.sh]: Please add at least a shebang. Ideally the script should also fail if something inside fails (even though taupage-init currently does nothing) so we can improve it later.

File: runtime/etc/init/td-agent.conf [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+stop on shutdown
+
+respawn
+
+pre-start script
+    mkdir -p -m0755 /var/run/td-agent
+end script
+
+exec /opt/td-agent/embedded/bin/ruby -Eascii-8bit:ascii-8bit /usr/sbin/td-agent --log /var/log/td-agent/td-agent.log --daemon /var/run/td-agent/td-agent.pid --no-supervisor
```
Comment (aermakov-zalando) [inline runtime/etc/init/td-agent.conf]: Who creates `/var/log/td-agent`?

File: runtime/opt/taupage/init.d/02-prepare-td-agent.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+# Ensure that directories are writable
+chmod -R 1777 /tmp/
```
Comment (aermakov-zalando) [inline runtime/opt/taupage/init.d/02-prepare-td-agent.sh]: Again, why?

File: runtime/opt/taupage/init.d/02-register-td-agent.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        # If scalyr_api_key starts with "aws:kms:" then decrypt key
+        match_kms_key = re.search('aws:kms:', scalyr_api_key, re.IGNORECASE)
+        if match_kms_key:
+            scalyr_api_key = re.sub(r'aws:kms:', '', scalyr_api_key)
+            try:
+                scalyr_api_key = subprocess.check_output(['python3',
+                                                          '/opt/taupage/bin/decrypt-kms.py',
+                                                          scalyr_api_key]).decode('UTF-8').strip()
+            except Exception:
+                logger.error('Failed to run /opt/taupage/bin/decrypt-kms.py')
```
Comment (aermakov-zalando) [inline runtime/opt/taupage/init.d/02-register-td-agent.py]: Why not just crash here? What's the point of proceeding?

File: build/setup.d/20-install-packages.sh [hunk 6]
```diff
@@ -33,7 +33,9 @@ rkhunter
 rsyslog-gnutls
 ruby
 scalyr-agent-2
+supervisor
```
Comment (aermakov-zalando) [inline build/setup.d/20-install-packages.sh]: Why? It doesn't seem to be used.

File: runtime/opt/taupage/init.d/02-prepare-td-agent.sh [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Comment (gargravarr) [inline runtime/opt/taupage/init.d/02-prepare-td-agent.sh]: #!/usr/bin/env bash

File: runtime/opt/taupage/init.d/02-register-td-agent.py (hunk 5 above)
Comment (vwiessner) [reply runtime/opt/taupage/init.d/02-register-td-agent.py]: There could be addititional destinations whose configuration depends on this script finishing, which should still work if the Scalyr key decryption fails.

File: runtime/opt/taupage/init.d/02-prepare-td-agent.sh (hunk 7 above)
Comment (femueller) [reply runtime/opt/taupage/init.d/02-prepare-td-agent.sh]: Why? We use `#!/bin/bash` for most scripts.

File: build/setup.d/21-install-td-agent-packages.sh (hunk 1 above)
Comment (femueller) [reply build/setup.d/21-install-td-agent-packages.sh]: I can't remember the exact reason. I will revert this change and see if we still require it.

File: runtime/opt/taupage/init.d/02-prepare-td-agent.sh (hunk 4 above)
Comment (femueller) [reply runtime/opt/taupage/init.d/02-prepare-td-agent.sh]: Same as above https://github.com/zalando-stups/taupage/pull/511#discussion_r220101407.

File: build/setup.d/21-install-td-agent-packages.sh (hunk 2 above)
Comment (femueller) [reply build/setup.d/21-install-td-agent-packages.sh]: Will do.

File: build/setup.d/21-install-td-agent-packages.sh (hunk 1 above)
Comment (femueller) [reply build/setup.d/21-install-td-agent-packages.sh]: This doesn't appear to be necessary.

File: runtime/opt/taupage/init.d/02-prepare-td-agent.sh (hunk 4 above)
Comment (femueller) [reply runtime/opt/taupage/init.d/02-prepare-td-agent.sh]: This doesn't appear to be necessary anymore.

File: runtime/opt/taupage/init.d/02-register-td-agent.py (hunk 5 above)
Comment (aermakov-zalando) [reply runtime/opt/taupage/init.d/02-register-td-agent.py]: But then the config will be invalid since you don't quote the api key in the template. If you want to rely on this, please test the behavior.

File: build/setup.d/21-install-td-agent-packages.sh (hunk 2 above)
Comment (aermakov-zalando) [reply build/setup.d/21-install-td-agent-packages.sh]: The script still doesn't fail.

**Comment by femueller** [comment]
@vwiessner and me have submitted the requested changes. Can you please check again?

**Comment by aermakov-zalando** [comment]
👍

**Comment by vwiessner** [comment]
:+1:

**Comment by aermakov-zalando** [comment]
👎

**Comment by aermakov-zalando** [comment]
Oops. I've just noticed that you fixed Travis in the PR as well. Do you mind removing the fix from here since it's fixed in `master`?

**Comment by vwiessner** [comment]
:+1:

**Comment by aermakov-zalando** [comment]
👍
